version: '3.8'

services:

  db:
    image: postgres:15.1-alpine
    restart: on-failure
    environment:
      POSTGRES_PASSWORD: ${POSTGRES_PASSWORD}
      POSTGRES_USER: ${POSTGRES_USER}
      POSTGRES_DB: ${POSTGRES_DB}
    volumes:
      - ./postgres_data:/var/lib/postgresql/data
    ports:
      - '${POSTGRES_PORT:-5432}:5432'
    healthcheck:
      test: pg_isready -U ${POSTGRES_USER} -d ${POSTGRES_DB}
      interval: 3s
      timeout: 3s
      retries: 3

  migrations:
    build: .
    depends_on:
      db:
        condition: service_healthy
    volumes:
      - ./.env:/todolist/.env
    command: python manage.py migrate

  api:
    build: .
    restart: on-failure
    ports:
      - '${API_PORT:-8000}:8000'
    volumes:
      - ./:/todolist/
    depends_on:
      db:
        condition: service_healthy
      migrations:
        condition: service_completed_successfully

  bot:
    build: .
    depends_on:
      db:
        condition: service_healthy
      migrations:
        condition: service_completed_successfully
    volumes:
      - ./.env:/todolist/.env
    command: python manage.py runbot

  frontend:
    image: sermalenk/skypro-front:lesson-38
    restart: on-failure
    ports:
      - '${FRONTEND_PORT:-80}:80'
    depends_on:
      - api
